elm/wizard.py

```diff
@@ -579,19 +586,28 @@ def query_vector_db(self, query, probes=25, limit=100):
 
         query_embedding = self.get_embedding(query)
 
-        self.cursor.execute(f"SET LOCAL ivfflat.probes = {probes};"
-                            f"SELECT {self.db_table}.id, "
-                            f"{self.db_table}.chunks, "
-                            f"{self.db_table}.embedding "
-                            "<=> %s::vector as score, "
-                            f"{self.db_table}.title, "
-                            f"{self.db_table}.authors, "
-                            f"{self.db_table}.year "
-                            f"FROM {self.db_schema}.{self.db_table} "
-                            "ORDER BY embedding <=> %s::vector LIMIT %s;",
-                            (query_embedding, query_embedding, limit,), )
-
-        result = self.cursor.fetchall()
+        with self.psycopg2.connect(**self.db_kwargs) as conn:
+            cursor = conn.cursor()
```

Comment:
yeah, this one is to address some database errors. The short story is that when users submitted queries without letting the previous one finish it would break the cursor and then the app could not generate a response. Here we are creating a cursor for each question to avoid the error and implementing a rollback to ensure that one error doesn't prevent the user from asking additional questions. I left `self.cursor` in the code for the sake of utility so that we can access other parts of the db through an `EnergyWizardPostgres` object during validation and such.

Reply:
That's certainly surprising! I would have guessed something like that should be addressed on the frontend (e.g. in your example, what would the current version of the code do with both outputs?). But oh well!

I think you should definitely keep it as is then, but definitely add some documentation around it (e.g. in the docstring or as a comment above this line) to explain the unexpected behavior. That way someone doesn't accidentally revert this change (it's untested, so someone can unknowingly re-introduce the bug).

Reply:
good idea!